Security and Privacy section

igrigorik opened the pull request:
Preview: https://rawgit.com/w3c/resource-hints/security/index.html#security-and-privacy
- explain confidentiality risks
- explain interaction with CSP, Mixed Content, and Referrer
- explain site author best-practices

igrigorik committed:
first cut at security and privacy section

- explain confidentiality risks
- explain interaction with CSP, Mixed Content, and Referrer
- explain site author best-practices

yoavweiss commented on index.html:
```diff
@@ -245,6 +245,28 @@
   <section>
+    <h2>Security and Privacy</h2>
+
+    <p>The use of <a href="#dfn-resource-hint-link">resource hints</a> hints can expose new information about both the current page and user's activity on it. For example, observing the content of prefetched or prerendered resources, or observing the DNS, IP, and hostname information exposed during connection establishment, both for secure and plain-text connections, can reveal information about  ...
```
I'm not sure I understand what new information is exposed here (that cannot already be explicitly exposed via JS and XHRs).

igrigorik replied:
It's not that it's new, it's more about documenting the surface area, associated risks, and providing recommendations for how to account for these things... both on the UA implementation side, and for the site author.

yoavweiss replied:
Then maybe s/can expose new information/can provide new means to expose information/

igrigorik replied:
Makes sense, updated in 51be2cb - thanks!

yoavweiss commented on index.html:
```diff
@@ -245,6 +245,28 @@
+    <ul>
+      <li>Treat <a>resource hint link</a>'s as <a href="https://w3c.github.io/webappsec/specs/mixedcontent/#category-optionally-blockable">Optionally-blockable Content</a>, as defined by [[MIXED-CONTENT]].</li>
+      <li>Enforce CSP policy restrictions for <a>resource hint link</a>'s, as defined by [[CSP2]].</li>
+      <li>Apply relevant the <a href="https://w3c.github.io/webappsec/specs/CSP2/#complications">Processing Complications recommendations</a> for optimistic resource fetching, as defined by [[CSP2]].</li>
```
s/the//

yoavweiss commented:
Looks good, other than the fact that I don't get what new info is exposed that can't be exposed by other means, so it'd be good to give concrete examples.

@mikewest - can you take a look?

igrigorik committed:
fix typo

igrigorik committed:
RH can provide *new means* to expose information

RH does not expose new information.

yoavweiss commented:
LGTM.

I guess we can land and ask the security folks for review later on.

igrigorik commented:
@toddreifsteck @marcoscaceres can you guys please take a look at this one?

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+
+    <p>The use of <a href="#dfn-resource-hint-link">resource hints</a> can provide new means to expose information about both the current page and user's activity on it. For example, observing the content of prefetched or prerendered resources, or observing the DNS, IP, and hostname information exposed during connection establishment, both for secure and plain-text connections, can reveal informa ...
+
+    <p>To mitigate some of the above risks, the user agent MAY provide optional user setting to control the use of <a href="#dfn-resource-hint-link">resource hints</a>, and the user agent MUST:</p>
```
MAY and "optional" is tautological :) Drop the optional.

marcoscaceres replied:
I think you want a full stop where the last comma is. It's a bit confusing you go from a MAY to a MUST.

igrigorik replied:
Ack.

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+
+    <ul>
+      <li>Treat <a>resource hint link</a>'s as <a href="https://w3c.github.io/webappsec/specs/mixedcontent/#category-optionally-blockable">Optionally-blockable Content</a>, as defined by [[MIXED-CONTENT]].</li>
+      <li>Enforce CSP policy restrictions for <a>resource hint link</a>'s, as defined by [[CSP2]].</li>
```
s/[[MIXED-CONTENT]]/[[!MIXED-CONTENT]]

marcoscaceres replied:
s/[[CSP2]]/[[!CSP3]]

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+      <li>Treat <a>resource hint link</a>'s as <a href="https://w3c.github.io/webappsec/specs/mixedcontent/#category-optionally-blockable">Optionally-blockable Content</a>, as defined by [[MIXED-CONTENT]].</li>
+      <li>Enforce CSP policy restrictions for <a>resource hint link</a>'s, as defined by [[CSP2]].</li>
+      <li>Apply relevant <a href="https://w3c.github.io/webappsec/specs/CSP2/#complications">Processing Complications recommendations</a> for optimistic resource fetching, as defined by [[CSP2]].</li>
+      <li>Apply relevant Referrer policy restrictions for <a>resource hint link</a>'s, as defined by [[REFERRER-POLICY]].</li>
```
As above, I think you should reference CSP3 instead.

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+    <ul>
+      <li>Treat <a>resource hint link</a>'s as <a href="https://w3c.github.io/webappsec/specs/mixedcontent/#category-optionally-blockable">Optionally-blockable Content</a>, as defined by [[MIXED-CONTENT]].</li>
+      <li>Enforce CSP policy restrictions for <a>resource hint link</a>'s, as defined by [[CSP2]].</li>
+      <li>Apply relevant <a href="https://w3c.github.io/webappsec/specs/CSP2/#complications">Processing Complications recommendations</a> for optimistic resource fetching, as defined by [[CSP2]].</li>
```
As above, I think you should reference CSP3 instead.

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+      <li>Apply relevant <a href="https://w3c.github.io/webappsec/specs/CSP2/#complications">Processing Complications recommendations</a> for optimistic resource fetching, as defined by [[CSP2]].</li>
+      <li>Apply relevant Referrer policy restrictions for <a>resource hint link</a>'s, as defined by [[REFERRER-POLICY]].</li>
+    </ul>
+
```
Make sure the links to specs in this section have a `[[!` to make them normative.

igrigorik replied:
Ack.

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+      <li>Apply relevant Referrer policy restrictions for <a>resource hint link</a>'s, as defined by [[REFERRER-POLICY]].</li>
+    </ul>
+
+    <p>Additionally, the site author SHOULD take necessary precautions to specify the relevant [[MIXED-CONTENT]], [[CSP]], and [[REFERRER-POLICY]] policies to allow the user agent to register and apply them prior to processing a <a>resource hint link</a>:</p>
```
Conformance requirements for authors are not great. Maybe say "Additionally, site authors needs to take..."

marcoscaceres replied:
I think you need to give more clear guidance to authors here as to how they do the above.

marcoscaceres replied:
(actually, ignore the above).

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+    <p>Additionally, the site author SHOULD take necessary precautions to specify the relevant [[MIXED-CONTENT]], [[CSP]], and [[REFERRER-POLICY]] policies to allow the user agent to register and apply them prior to processing a <a>resource hint link</a>:</p>
+
+    <ul>
+      <li>If <a>resource hint link</a> is provided via HTML markup, then the relevant `<meta>` policies should be declared ahead of them - e.g. <a href="http://www.w3.org/TR/CSP2/#delivery-html-meta-element">see `<meta>` recommendations</a> for CSP.</li>
```
should/SHOULD ?

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+
+    <ul>
+      <li>If <a>resource hint link</a> is provided via HTML markup, then the relevant `<meta>` policies should be declared ahead of them - e.g. <a href="http://www.w3.org/TR/CSP2/#delivery-html-meta-element">see `<meta>` recommendations</a> for CSP.</li>
+      <li>If <a>resource hint link</a> is provided via the `Link` HTTP response header, then the relevant policies should also be delivered as an HTTP response header - e.g. see <a href="https://w3c.github.io/webappsec/specs/CSP2/#complications">Processing Complications</a> for CSP.</li>
```
Point to CSP3

marcoscaceres commented on index.html:
```diff
@@ -245,6 +245,28 @@
+    <h2>Security and Privacy</h2>
+
+    <p>The use of <a href="#dfn-resource-hint-link">resource hints</a> can provide new means to expose information about both the current page and user's activity on it. For example, observing the content of prefetched or prerendered resources, or observing the DNS, IP, and hostname information exposed during connection establishment, both for secure and plain-text connections, can reveal informa ...
+
```
This sounds kinda serious. Maybe a candidate for use only in TLS environments?

igrigorik replied:
Note the "new means"... all of this is possible today by simply observing what resources the site is fetching and from which origins. I just want to be thorough in documenting the possible risks.

igrigorik committed:
cleanup privacy section

- split MAY/MUST language
- normative references for spec links
- reference CSP3

igrigorik commented:
@marcoscaceres thanks for reviewing it! Updated, ptal.

igrigorik committed:
shuffle use cases and security

dropping CSP clause because CSP does not definity any policies for
controlling prefetch, prerender, and preconnect is out of scope.
